### ISSUE TITLE
" $fragmentRefs" is unsafe in TS 3.6+

### ISSUE THREAD

**Issue by ds300** [relay-tools/relay-compiler-language-typescript#138]
The technique we are using to implement composable opaque types no longer works in TS 3.6+

The problem is that the intersection of unique symbol types now resolves to `never`

```ts
declare const FragmentA$ref: unique symbol
type FragmentA$ref = typeof FragmentA$ref
declare const FragmentB$ref: unique symbol
type FragmentB$ref = typeof FragmentB$ref

type CombinedRefs = FragmentA$ref & FragmentB$ref
// CombinedRefs = never
```

`never` is assignable to all other types, which means that any query or fragment data which uses more than one fragment spread is now assignable to any other 

```ts
declare const FragmentC$ref: unique symbol
type FragmentC$ref = typeof FragmentC$ref

type SelectionAB = {
    " $fragmentRefs": FragmentA$ref & FragmentB$ref
}

type SelectionC = {
    " $fragmentRefs": FragmentC$ref
}

declare const ab: SelectionAB
const c: SelectionC = ab // TS does not complain :(
```

I'd propose switching to something like the following:

```ts
type FragmentRefs<T extends symbol> =  null | {[k in T]: true}

type SafeSelectionAB = {
    " $fragmentRefs": FragmentRefs<FragmentA$ref | FragmentB$ref>
}

type SafeSelectionC = {
    " $fragmentRefs": FragmentRefs<FragmentC$ref>
}

declare const sab: SafeSelectionAB
const sc: SafeSelectionC = sab

// Type 'SafeSelectionAB' is not assignable to type 'SafeSelectionC'.
//  Types of property '" $fragmentRefs"' are incompatible.
//    Type 'FragmentRefs<unique symbol | unique symbol>' is not assignable to type 'FragmentRefs<unique symbol>'.
//      Property '[FragmentC$ref]' is missing in type '{ [FragmentA$ref]: true; [FragmentB$ref]: true; }' but required in type '{ [FragmentC$ref]: true; }'.
```

We could even switch to using string literals based on the fragment name for better error messages, at the loss of a tiny tiny bit of safety.

```ts
type FragmentRefs<T extends string> =  null | {[k in T]: true}

type SafeSelectionAB = {
    " $fragmentRefs": FragmentRefs<"FragmentA" | "FragmentB">
}

type SafeSelectionC = {
    " $fragmentRefs": FragmentRefs<"FragmentC">
}

declare const sab: SafeSelectionAB
const sc: SafeSelectionC = sab

// Type 'SafeSelectionAB' is not assignable to type 'SafeSelectionC'.
//  Types of property '" $fragmentRefs"' are incompatible.
//    Type 'FragmentRefs<"FragmentA" | "FragmentB">' is not assignable to type 'FragmentRefs<"FragmentC">'.
//      Property 'FragmentC' is missing in type '{ FragmentA: true; FragmentB: true; }' but required in type '{ FragmentC: true; }'.
```

**Comment by alloy** [comment]
I like that idea of using strings and improving DX that way. Can’t we just use a plain string union in that case, rather than creating an object?

**Comment by alloy** [comment]
Btw, don’t you mean in TS 3.6+ ?

**Comment by ds300** [comment]
Yes I meant 3.6, updated, thanks!

And no plain string unions won't work because `"a" | "b"` is not assignable to `"a"`, but objects will work because `{a: true, b: true}` is assignable to `{a: true}`

**Comment by alloy** [comment]
Ah yep, makes sense 👍 Want to take a stab at creating a PR for that?

**Comment by ds300** [comment]
Sure

**Comment by ds300** [comment]
How do you feel about using string literals for the `" $refType"` property too? Would let us delete a fair amount of code.

**Comment by alloy** [comment]
A little hard to imagine the changes from the top of my mind, but I’m always in favor of simplification.

(cc @kastermester for visibility)

**Comment by ds300** [comment]
This was fixed in #139
